### PR TITLE
Spline #191 Add Expressions in details panel

### PR DESCRIPTION
**Pull request by leborgneromain**
Spline #191 Add expressions to details panel

### CODE REVIEW

File: client-ui/src/app/components/lineage/lineage-details/schema-details/schema-details.component.ts [hunk 1]
```diff
@@ -76,103 +95,55 @@ export class SchemaDetailsComponent implements AfterViewInit {
...
+          diff.map(item => expressions.push(new Expression("Dropped Attributes", item.name, null)))
         }
         break
-      //TODO : Implement the other expressions for the other types
+      case OperationType.Aggregate:
+        properties.aggregateExpressions.forEach(aggregate => expressions.push(new Expression("Aggregate", getText(aggregate[1], attributeList), aggregate[1])))
+        properties.groupingExpressions.forEach(grouping => expressions.push(new Expression("Grouping", getText(grouping[1], attributeList), grouping[1])))
+        break
+      case OperationType.LogicalRelation:
+        const inputSourceExpression = new Expression("inputSource", properties.inputSources[0], null)
```
Comment (wajda) [inline client-ui/src/app/components/lineage/lineage-details/schema-details/schema-details.component.ts]: Why is `inputSource` expression? And why is only first one taken?

File: client-ui/src/app/components/lineage/lineage-details/schema-details/projection/projection.component.ts [hunk 2]
```diff
@@ -13,12 +13,19 @@
...
 @Component({
   selector: 'projection',
   templateUrl: './projection.component.html',
   styleUrls: ['./projection.component.less']
 })
-export class ProjectionComponent extends ExpressionComponent { }
+export class ProjectionComponent extends ExpressionComponent {
+
+  public hasDroppedAttribute(expressions: Expression[]): boolean {
+    return expressions.filter(e => e.title == "Dropped Attributes").length > 0
```
Comment (wajda) [inline client-ui/src/app/components/lineage/lineage-details/schema-details/projection/projection.component.ts]: Why do you use `title` property as type? Shouldn't enumeration be used for this purpose instead?

File: client-ui/src/app/components/lineage/lineage-details/schema-details/aggregate/aggregate.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+
+import { AggregateComponent } from './aggregate.component';
+
+describe('AggregateComponent', () => {
+  let component: AggregateComponent;
```
[the remark this answers is not in the document]
Comment (leborgneromain) [reply client-ui/src/app/components/lineage/lineage-details/schema-details/aggregate/aggregate.component.spec.ts]: The code is generated from Angular CLI

File: client-ui/src/app/components/lineage/lineage-details/schema-details/aggregate/aggregate.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  let fixture: ComponentFixture<AggregateComponent>;
+
+  beforeEach(async(() => {
+    TestBed.configureTestingModule({
+      declarations: [AggregateComponent]
+    })
+      .compileComponents();
+  }));
```
[the remark this answers is not in the document]
Comment (leborgneromain) [reply client-ui/src/app/components/lineage/lineage-details/schema-details/aggregate/aggregate.component.spec.ts]: The code is generated from Angular CLI

File: spark/agent/src/main/scala/za/co/absa/spline/harvester/json/SplineShortTypeHints.scala [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package za.co.absa.spline.harvester.json
+
+import org.json4s.TypeHints
+
+case class SplineShortTypeHints(hints: List[Class[_]]) extends TypeHints {
```
[the remark this answers is not in the document]
Comment (leborgneromain) [reply spark/agent/src/main/scala/za/co/absa/spline/harvester/json/SplineShortTypeHints.scala]: I've created from ShortTypeHints class which is a case class

File: client-ui/src/app/components/lineage/lineage-details/schema-details/schema-details.component.ts [hunk 6]
```diff
@@ -76,103 +95,55 @@ export class SchemaDetailsComponent implements AfterViewInit {
...
-          const title = "Dropped Attributes"
-          const values = diff.map(item => item.name);
-          const droppedAttributes = new Expression(title, values)
-          expressions.push(droppedAttributes)
+          diff.map(item => expressions.push(new Expression("Dropped Attributes", item.name, null)))
         }
         break
-      //TODO : Implement the other expressions for the other types
+      case OperationType.Aggregate:
+        properties.aggregateExpressions.forEach(aggregate => expressions.push(new Expression("Aggregate", getText(aggregate[1], attributeList), aggregate[1])))
```
[the remark this answers is not in the document]
Comment (leborgneromain) [reply client-ui/src/app/components/lineage/lineage-details/schema-details/schema-details.component.ts]: done

File: client-ui/src/app/components/lineage/lineage-details/schema-details/schema-details.component.ts [hunk 7]
```diff
@@ -76,103 +95,55 @@ export class SchemaDetailsComponent implements AfterViewInit {
...
-        const output = attribute.schemas[attribute.output]
+        _.each(detailsInfos.inputs, schemaIndex => inputs = _.concat(inputs, detailsInfos.schemas[schemaIndex]))
+        const output = detailsInfos.schemas[detailsInfos.output]
         const diff = _.differenceBy(inputs, output, 'name')
         if (diff.length > 0) {
-          const title = "Dropped Attributes"
-          const values = diff.map(item => item.name);
-          const droppedAttributes = new Expression(title, values)
-          expressions.push(droppedAttributes)
+          diff.map(item => expressions.push(new Expression("Dropped Attributes", item.name, null)))
```
[the remark this answers is not in the document]
Comment (leborgneromain) [reply client-ui/src/app/components/lineage/lineage-details/schema-details/schema-details.component.ts]: fixed

File: client-ui/src/app/components/lineage/lineage-details/schema-details/schema-details.component.ts (hunk 1 above)
Comment (leborgneromain) [reply client-ui/src/app/components/lineage/lineage-details/schema-details/schema-details.component.ts]: fixed

File: client-ui/src/app/components/lineage/lineage-details/schema-details/projection/projection.component.ts (hunk 2 above)
Comment (leborgneromain) [reply client-ui/src/app/components/lineage/lineage-details/schema-details/projection/projection.component.ts]: fixed

File: client-ui/src/app/store/reducers/expression.reducer.ts [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { IExpression, IBinary, IUDF, IAttrRef, IAlias, IGenericLeaf, IGeneric, ILiteral } from 'src/app/model/expression-model'
+import { ExpressionType } from 'src/app/model/types/expressionType'
+import * as _ from 'lodash'
+
+export function getName(expr: IExpression, attributeList: any): string {
```
[the remark this answers is not in the document]
Comment (leborgneromain) [reply client-ui/src/app/store/reducers/expression.reducer.ts]: Created a directory "util" with all the global functions that are not related to store but used at different places in the app

File: client-ui/src/app/components/lineage/lineage-details/schema-details/schema-details.component.ts [hunk 9]
```diff
@@ -44,23 +43,43 @@ export class SchemaDetailsComponent implements AfterViewInit {
...
+                return this.store
+                  .select('executedLogicalPlan')
+                  .pipe(
+                    map(executedLogicalPlan => {
+                      return { detailsInfos: detailsInfos, executedLogicalPlan: executedLogicalPlan }
+                    })
+                  )
+              })
+            )
+        )
```
[the remark this answers is not in the document]
Comment (leborgneromain) [reply client-ui/src/app/components/lineage/lineage-details/schema-details/schema-details.component.ts]: reduced a bit that pyramid

File: spark/agent/src/main/scala/za/co/absa/spline/harvester/json/SplineShortTypeHints.scala [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+  def hintFor(clazz: Class[_]): String = {
+    val classNameSplit : Array[String] = clazz.getName.split('.')
+    s"${classNameSplit(classNameSplit.length-2)}.${classNameSplit(classNameSplit.length-1)}"
+  }
```
[the remark this answers is not in the document]
Comment (leborgneromain) [reply spark/agent/src/main/scala/za/co/absa/spline/harvester/json/SplineShortTypeHints.scala]: thanks

[the remark this answers is not in the document]
Comment (leborgneromain) [reply spark/agent/src/main/scala/za/co/absa/spline/harvester/json/SplineShortTypeHints.scala]: fixed

**Pull request merged**